File: Backend/NootsWEB/Noots.Editor/Entities/Text/BlockDiff.cs
using Common.DatabaseModels.Models.NoteContent.TextContent.TextBlockElements;
using Noots.RGA_CRDT;

namespace Noots.Editor.Entities.Text;

public class BlockDiff
{
    public TreeRGA<string> Tree { set; get; } // Text

    public BlockProperty<string> HighlightColor { set; get; } // highlightColor

    public BlockProperty<string> TextColor { set; get; } // textColor

    public BlockProperty<string> Link { set; get; } // link

    public BlockProperty<List<TextType>> TextTypes { set; get; } // textTypes

    public MergeTransaction<string> MergeOps { set; get; }
}

File: Client/src/app/content/notes/full-note/content-editor-services/models/block-diff.ts
import { Letter } from '../../../models/editor-models/text-models/letter';
import { TextType } from '../../../models/editor-models/text-models/text-type';

export class BlockDiff {
  highlightColor?: string; // this field can be null and it`s ok, null = no background;

  textColor?: string; // this field can be null and it`s ok, null = white;

  link?: string;

  textTypes?: TextType[];

  readonly letterIdsToDelete?: string[] = [];

  readonly lettersToAdd?: Letter[] = [];

  constructor(public id: string) {}

  get isAnyChanges(): boolean {
    const flag =
      this.highlightColor !== undefined ||
      this.textColor !== undefined ||
      this.link !== undefined ||
      this.textTypes !== undefined ||
      this.letterIdsToDelete?.length > 0 ||
      this.lettersToAdd?.length > 0;
    return flag;
  }

  init(block: Partial<BlockDiff>): void {
    this.highlightColor = block.highlightColor === 'd' ? null : block.highlightColor;
    this.textColor = block.textColor === 'd' ? null : block.textColor;
    this.link = block.link;
    this.textTypes = block.textTypes;

    this.letterIdsToDelete.push(...block.letterIdsToDelete);
    this.lettersToAdd.push(
      ...block.lettersToAdd.map((x) => new Letter(x.symbol, x.fractionalIndex, x.id)),
    );
  }
}
